File: core/src/main/java/com/views/textview/color/LColorTextView.java
package com.views.textview.color;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.Html;
import android.text.TextUtils;
import android.util.AttributeSet;

import androidx.appcompat.widget.AppCompatTextView;

import com.R;

import java.util.ArrayList;
import java.util.List;
//TODO convert kotlin
public class LColorTextView extends AppCompatTextView {
    private List<String> mColorTexts = new ArrayList<>();
    private List<String> mColors = new ArrayList<>();
    private int size;
    private String mCurrentText;

    public LColorTextView(Context context) {
        super(context);
    }

    public LColorTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public LColorTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    private void init(AttributeSet attrs) {
        final TypedArray attrsArray = getContext().obtainStyledAttributes(attrs, R.styleable.LColorTextView, 0, 0);
        final String colorTexts = attrsArray.getString(R.styleable.LColorTextView_color_texts);
        final String colors = attrsArray.getString(R.styleable.LColorTextView_color_arrays);
        attrsArray.recycle();

        initData(colorTexts, colors);

        setHtmlText();
    }

    private void initData(final String colorTexts, final String colors) {
        try {
            if (!TextUtils.isEmpty(colorTexts)) {
                String[] texts = colorTexts.split("\\|");
                for (int i = 0; i < texts.length; i++) {
                    mColorTexts.add(texts[i]);
                }
            }

            if (!TextUtils.isEmpty(colors)) {
                String[] texts = colors.split("\\|");
                for (int i = 0; i < texts.length; i++) {
                    mColors.add(texts[i]);
                }
            }
            size = Math.min(mColors.size(), mColorTexts.size());
            mCurrentText = getText().toString();
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }

    private void setHtmlText() {
        try {
            if (!TextUtils.isEmpty(mCurrentText)) {
                for (int i = 0; i < size; i++) {
                    mCurrentText = mCurrentText.replace(mColorTexts.get(i), color(mColors.get(i), mColorTexts.get(i)));
                }
            }
            if (!TextUtils.isEmpty(mCurrentText)) {
                setText(Html.fromHtml(mCurrentText));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    public LColorTextView findAndSetStrColor(String str, String color) {
        try {
            if (!TextUtils.isEmpty(mCurrentText)) {
                mCurrentText = mCurrentText.replaceAll(str, color(color, str));
            }
            if (!TextUtils.isEmpty(mCurrentText)) {
                setText(Html.fromHtml(mCurrentText));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return this;
    }


    public String color(String colorCode, String str) {
        return "<font color=\"" + colorCode + "\">" + str + "</font>";
    }
}

File: core/src/main/java/com/loitpcore/views/textView/color/LColorTextView.kt
package com.loitpcore.views.textView.color

import android.content.Context
import android.text.Html
import android.text.TextUtils
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.loitpcore.R
import kotlin.math.min

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class LColorTextView : AppCompatTextView {
    private val mColorTexts: MutableList<String> = ArrayList()
    private val mColors: MutableList<String> = ArrayList()
    private var size = 0
    private var mCurrentText: String? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(attrs)
    }

    private fun init(attrs: AttributeSet?) {
        val attrsArray = context.obtainStyledAttributes(attrs, R.styleable.LColorTextView, 0, 0)
        val colorTexts = attrsArray.getString(R.styleable.LColorTextView_color_texts)
        val colors = attrsArray.getString(R.styleable.LColorTextView_color_arrays)
        attrsArray.recycle()
        initData(colorTexts, colors)
        setHtmlText()
    }

    private fun initData(colorTexts: String?, colors: String?) {
        try {
            if (!TextUtils.isEmpty(colorTexts)) {
                val texts = colorTexts?.split("\\|".toRegex())?.toTypedArray()
                texts?.let {
                    for (i in it.indices) {
                        mColorTexts.add(it[i])
                    }
                }
            }
            if (!TextUtils.isEmpty(colors)) {
                val texts = colors?.split("\\|".toRegex())?.toTypedArray()
                texts?.let {
                    for (i in it.indices) {
                        mColors.add(it[i])
                    }
                }
            }
            size = min(mColors.size, mColorTexts.size)
            mCurrentText = text.toString()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun setHtmlText() {
        try {
            if (!TextUtils.isEmpty(mCurrentText)) {
                for (i in 0 until size) {
                    mCurrentText =
                        mCurrentText?.replace(mColorTexts[i], color(mColors[i], mColorTexts[i]))
                }
            }
            if (!TextUtils.isEmpty(mCurrentText)) {
                text = Html.fromHtml(mCurrentText)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Suppress("unused")
    fun findAndSetStrColor(str: String, color: String): LColorTextView {
        try {
            if (!TextUtils.isEmpty(mCurrentText)) {
                mCurrentText = mCurrentText?.replace(str.toRegex(), color(color, str))
            }
            if (!TextUtils.isEmpty(mCurrentText)) {
                text = Html.fromHtml(mCurrentText)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return this
    }

    fun color(colorCode: String, str: String): String {
        return "<font color=\"$colorCode\">$str</font>"
    }
}
